app/src/main/java/com/keylesspalace/tusky/viewmodel/EditProfileViewModel.kt

```diff
@@ -96,29 +106,73 @@ class EditProfileViewModel @Inject constructor(
         headerData.value = getHeaderUri()
     }
 
-    fun save(newDisplayName: String, newNote: String, newLocked: Boolean, newFields: List<StringField>) {
+    internal fun save(newProfileData: ProfileData) {
         if (saveData.value is Loading || profileData.value !is Success) {
             return
         }
 
         saveData.value = Loading()
 
-        val displayName = if (oldProfileData?.displayName == newDisplayName) {
+        val encoded = encodeChangedProfileFields(newProfileData)
```

Comment:
Could this be done directly with one function `hasChangedFields(newProfileData)` - instead of `encode + allFieldsAreNull``?

(As it is not really encoded but only the diff is determined?)

Reply:
Not as-is, since the result of this "encoding" is used below; and the result of "allFieldsAreNull" is used separately as well... 

With some refactoring this could be reused, but it wasn't my intention to alter the whole EditProfile logic, as it's quite complex as it is. 

Do you have a concrete idea in mind?

Reply:
My concrete idea (haven't tried it, will do if this comment "does not work" :-)):

- Have a data class "ProfileData" with all fields relevant for a profile.
- Use that as data type / parameter.
- For "is it changed?": the view model should compare against this data model (with the new data)
- Can also be used in the api call (or its fields)

Reply:
Yes, that's actually what I had initially in mind, you could even compare the hash of the "old" vs. "new". 
Perhaps that could be done as either part of this PR or a new PR, because it will involve touching this VM for sure. 

I have created the ProfileData, in this PR. What initially caught me "off guard" was that the VM is "encoding" this and keeping *that* in memory, that's why I had to extract these encoding fields to compare apples to apples. 

I'd say all this encoding and what not should not be a VM problem, rather the Repo/Use case that deals with the persistence/storage/API/etc.

What do you think?

Reply:
I see:

The api expects a) only changed data AND b) in an encoded form.

So what I now did is separate the diff-making from the encoding. Mixing them leads to confusion (e. g. an oddly named data model).
Also I renamed some stuff to honor that.

See if you like it: https://github.com/Gryzor/Tusky/pull/1